chanotify/chanotify.go

```diff
@@ -17,85 +11,68 @@ type dataPair struct {
 // from Notifier.
 type Notifier struct {
-	c     chan string
-	chMap map[<-chan struct{}]*dataPair
-	ctrl  chan struct{}
-	m     sync.Mutex
+	c chan string
+
+	m      sync.Mutex // guards doneCh
+	doneCh map[string]chan struct{}
 }
 
-// New returns already running *Notifier.
+// New returns a new *Notifier.
 func New() *Notifier {
 	s := &Notifier{
-		c:     make(chan string),
-		chMap: make(map[<-chan struct{}]*dataPair),
-		ctrl:  make(chan struct{}),
+		c:      make(chan string),
+		doneCh: make(map[string]chan struct{}),
 	}
-	go s.start()
 	return s
 }
 
 // Chan returns channel on which client listen for notifications.
-// Ids of notifications is sent to that channel.
+// IDs of notifications is sent to the returned channel.
 func (s *Notifier) Chan() <-chan string {
 	return s.c
 }
 
-// Add adds new notification channel to Notifier.
-func (s *Notifier) Add(ch <-chan struct{}, id string) {
+func (s *Notifier) killWorker(id string, done chan struct{}) {
 	s.m.Lock()
-	s.chMap[ch] = &dataPair{
-		id: id,
-		selectCase: reflect.SelectCase{
-			Dir:  reflect.SelectRecv,
-			Chan: reflect.ValueOf(ch),
-		},
-	}
+	delete(s.doneCh, id)
 	s.m.Unlock()
-	s.ctrl <- struct{}{}
+	close(done)
 }
 
-// Close stops Notifier to listen for any notifications and closes its
-// "client-side" channel.
-func (s *Notifier) Close() {
-	close(s.ctrl)
-}
+// Add adds new notification channel to Notifier.
+func (s *Notifier) Add(ch <-chan struct{}, id string) {
+	done := make(chan struct{})
+	s.m.Lock()
+	s.doneCh[id] = done
+	s.m.Unlock()
 
-func (s *Notifier) start() {
-	for {
-		c := s.createCase()
-		i, _, ok := reflect.Select(c)
-		if i == 0 {
-			if ok {
-				continue
+	go func(ch <-chan struct{}, id string, done chan struct{}) {
+		for {
+			select {
+			case _, ok := <-ch:
+				if !ok {
+					// If the channel is closed, we don't need the goroutine
+					// or the done channel mechanism running anymore.
+					s.killWorker(id, done)
+					return
+				}
+				s.c <- id
+			case <-done:
+				// We don't need this goroutine running anymore, return.
+				s.killWorker(id, done)
+				return
 			}
-			// ctrl was closed, we can safely close output
-			close(s.c)
-			return
-		}
-		ch := c[i].Chan.Interface().(<-chan struct{})
-		if ok {
-			s.c <- s.chMap[ch].id
-			continue
 		}
-		// the channel was closed and we should remove it
-		s.m.Lock()
-		delete(s.chMap, ch)
-		s.m.Unlock()
-	}
+	}(ch, id, done)
 }
 
-func (s *Notifier) createCase() []reflect.SelectCase {
-	// put ctrl channel as 0 element of select
-	out := []reflect.SelectCase{
-		reflect.SelectCase{
-			Dir:  reflect.SelectRecv,
-			Chan: reflect.ValueOf(s.ctrl),
-		},
-	}
+// Close closes the notifier and releases its underlying resources.
+func (s *Notifier) Close() {
 	s.m.Lock()
-	for _, pair := range s.chMap {
-		out = append(out, pair.selectCase)
+	defer s.m.Unlock()
+	for _, done := range s.doneCh {
+		done <- struct{}{}
```

Comment:
I'd prefer to close those channels here. It's non-blocking and will have the same effect. But it's possible that I don't see something.